File: 0202-happy-number/0202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        int number = n; 
        unordered_map<int,int> hashMap; 
        while(number != 1){
            int happy = getNumber(number); 
            if(hashMap.count(happy)) return false; 

            hashMap[happy]++; 
            number = happy; 
        }
        return true; 
    }

    int getNumber(int number){
        int res = 0; 
        while(number){
            int lastDigit = number % 10; 
            res += lastDigit * lastDigit; 
            number /= 10; 
        }
        return res; 
    }
};

File: 0202-happy-number/0202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
        Map<Integer,Boolean> map = new HashMap<>(); 
        while(n > 1){
            int tmp = n; 
            int nextDigit = 0; 
            while(tmp > 0){
                int lastDigit = tmp % 10; 
                tmp /= 10; 
                nextDigit += lastDigit * lastDigit; 
            }
            n = nextDigit; 
            //System.out.println(n); 
            if(map.containsKey(n)) return false; 
            map.put(n,true); 
        }
        return true; 
    }
}